### ISSUE TITLE
An issue of the If Statement

### ISSUE THREAD

**Issue by YuanWangC** [chakra-core/ChakraCore#6377]
#### Description
When running the testcase below, a SyntaxError exception should be thrown while chakra does not.
According to ES10, the `Expression` in  parentheses of `IfStatement` should executed based on [this step](http://www.ecma-international.org/ecma-262/10.0/index.html#sec-comma-operator-runtime-semantics-evaluation). And a Comma-terminated expression obviously is not belong to valid `Expression`. This maybe an issue of Chakra.
#### Version
[eaaf7ac](https://github.com/microsoft/ChakraCore/commit/eaaf7ac6beca01ab0d80c05959f3868463043c2d)

#### Testcase
```
var NISLFuzzingFunc = function(){
    var f = 0;
    if(f === 0,) {
        print("run_here");
    }
}
NISLFuzzingFunc();
```
#### Ecpected output
```
 SyntaxError: Unexpected token ')'
```
#### Output
```
run_here
```

**Comment by rhuanjl** [comment]
Had a quick look - there's a missing check for a deferred comma error here: https://github.com/microsoft/ChakraCore/blob/409b0a2f1901d00992bd8f46a64bff6ed96b8cd2/lib/Parser/Parse.cpp#L10785

Should be an easy fix. (Not going to submit it right now but having read the issue I wanted to at least work out the cause so I or someone else can fix fairly quickly later).

**Comment by rhuanjl** [comment]
If anyone wants to try out contributing, this would be a very easy first fix to familiarise yourself with how to contribute to chakracore - I'll assist as needed if anyone picks this up.

**Comment by mum-never-proud** [comment]
I would love to work on this PR if no one has picked up yet :)

cc: @rhuanjl

**Comment by rhuanjl** [comment]
@mum-never-proud  It may be a little while until we can merge your work, as we're in the process of transitioning Chakracore from microsoft ownership to community ownership, but please give this a go and we'll hopefully be ready to merge things again in a week or so.

Let me know what help you need if any.

**Comment by rhuanjl** [comment]
@mum-never-proud If you're still interested we're now open to merging new commits - let me know if you need help with how to start.

**Comment by mum-never-proud** [comment]
@rhuanjl sure, I will raise a PR by today and will let you know if I am stuck

**Comment by mum-never-proud** [comment]
```c++
if (buildAST)
  {
     pnodeIf = CreateNodeForOpT<knopIf>(ichMin);
     pnodeIf->ichLim = this->GetScanner()->IchLimTok();
     pnodeIf->pnodeCond = pnodeCond;
  }
ChkCurTok(tkComma, ERRnoComma);
ChkCurTok(tkRParen, ERRnoRparen);
```
will the above check work?
I couldn't find a file where I can write a test case for if condition
and also in `perrors.h` there is no error message for a trailing comma, do i need to add it or is it somewhere else?
`ERRnoComma` is something I declared we can change to the suitable name

sorry if I am going in wrong direction thanks for your patience 💪 

cc @rhuanjl

**Comment by rhuanjl** [comment]
I don't think that's quite right, try just doing an if that looks at the value of `m_deferCommaError` I think it will be true when you need to error here.

You'll then want to use `this->GetScanner()->SeekTo(m_deferCommaErrorLoc);` (this sets the line/character back to where the comma was so it will show correctly in the error message)

For throwing the error you can use the `Error()` function - and either use ERRnoRparen as the error or add a new error to perrors.h. 

For testing this I figure a good place would be `test/Bugs/misc_bugs.js`

**For a bit more detail:**
1. `ChkCurTok` looks at the current character only and throws an error if it's not the requested one. So `ChkCurTok(tkRParen, ERRnoRparen);` is going to throw if the current character isn't a right Paren i.e. `)` 
2. Adding a ChkCurTok for a comma before the Right paren `)` one means you're going to throw if it's not a comma `,` AND then you're going to throw if it's not a right paren - this means you'll always throw.
3. In this case the ParesExpr instruction on line 10777 just above where you are will have parsed everything in the if's parentheses other than the right paren - so the parser is currently looking at that paren - you're one step passed the unwanted comma - so if you want to examine it you'd have to look back a step (which you could with GetScanner()->GetPrevious() BUT it shouldn't be necessary because I think ParseExpr should have set `m_deferCommaError` to true in the error case so you can just look at that.

**Comment by ylukomskyi** [comment]
Hello, I would like to try fixing this issue. Is anyone working on it at the moment?

**Comment by rhuanjl** [comment]
> Hello, I would like to try fixing this issue. Is anyone working on it at the moment?

It seems to be free, give it a go.

**Comment by ylukomskyi** [comment]
@rhuanjl I am having an issue with testing. I have extended `test/Bugs/misc_bugs.js` with the following test case
```
  {
    name: "should throw SyntaxError when expression within If is comma-terminated",
    body: function () {
        var func = function () {
          var f = 0;
          if (f === 0,) {
            print("run_here");
          }
        };
        assert.throws(() => func(), SyntaxError);
      }
  }
```
But when I am trying to run it the following error appears:
```
SyntaxError: Expected ')'
        at code (misc_bugs.js:308:16)
```
Any ideas regarding where I could add this test or make it work?

**Comment by rhuanjl** [comment]
To test syntax errors you have to use eval, you'll want to adjust your test to assert.throws that eval("snippet here") throws.

**Comment by rhuanjl** [comment]
Fixed via #6551 merged as #6580
